Placeholder fields and Osquery UI updates (#3045)
elastic/security-docs: docs/osquery/osquery-placeholder-fields.asciidoc

[[osquery-placeholder-fields]]
== Use placeholder fields in Osquery queries 

> Placeholder fields dynamically inject alert and event data into queries that you run. To create a placeholder field, you must wrap the field in double curly brackets (`{{}}`); you can use any field in an event or alert document. 

Author: ```suggestion
Placeholder fields dynamically inject alert and event data into Osquery queries at runtime. Placeholder fields use mustache syntax and must be wrapped in double curly brackets (`{{}}`); you can turn any field in an event or alert document into a placeholder field. 
```

[[osquery-placeholder-fields]]
== Use placeholder fields in Osquery queries 

> Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like parameters. You can use placeholder fields to build flexible and re-usable queries. 

Reviewer: ```suggestion
Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like parameters. You can use placeholder fields to build flexible and reusable queries. 
```

Placeholder fields work in single queries or query packs. They're also supported in the following features:

* <<alerts-run-osquery,Live queries>>
* <<osquery-response-action,Osquery Response Actions>>
+
* <<invest-guide-run-osquery,Investigation guides using Osquery queries>> 

[float]
[[placeholder-field-syntax]]
=== Placeholder field syntax and requirements

Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{example.field}}`). You can use any field within an event or alert document as a placeholder field. 

> Queries with placeholder fields can only run against alerts or events. Otherwise, they will lack the necessary values and be `Skipped`. 

Author: ```suggestion
Queries with placeholder fields can only run against alerts or events. Otherwise, they will lack the necessary values and the query status will be `error`. 
```

[float]
[[placeholder-field-example]]
==== Example query with a placeholder field

The following query uses the `{{host.name}}` placeholder field: 

> `SELECT * FROM os_version WHERE name = `{{host.name}}``

Author: ```suggestion
`SELECT * FROM os_version WHERE name = `{{host.os.name}}``
```

> When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.name}}` placeholder field. 

Reviewer: `passed into` sounds a little strange. Would `transferred` work here? 
```suggestion
When you run the query, the value that's stored in the alert or event's `host.name` field will be transferred to the `{{host.name}}` placeholder field. 
```

> When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.os.name}}` placeholder field. 

Author: ```suggestion
When you run the query, the value that's stored in the alert or event's `host.name` field will be transferred to the `{{host.os.name}}` placeholder field. 
```

[[osquery-placeholder-fields]]
== Use placeholder fields in Osquery queries 

> Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like replaceable parameters. You can use placeholder fields to build flexible and re-usable queries. 

Reviewer 2: I think maybe it would make more sense without the word "replaceable". I think it's redundant because a parameter is a variable that helps define some kind of function (in this case a query), so the name of the parameter in the query definition is meant to be replaced by some value, which I think is what this feature does. (I.e. you follow the syntax explained below to define how the parameter gets its value, and then the query runs with that value. )

> Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like replaceable parameters. You can use placeholder fields to build flexible and re-usable queries. 

Author: ```suggestion
Instead of hard-coding alert and event values into Osquery queries, you can use placeholder fields to dynamically pass this data into queries. Placeholder fields function like parameters. You can use placeholder fields to build flexible and re-usable queries. 
```

Placeholder fields work in single queries or query packs. They are also supported in the following features:

* <<alerts-run-osquery,Live queries>>

> * <<osquery-response-action,Osquery Response Action>>

Reviewer 2: ```suggestion
* <<osquery-response-action,Osquery Response Actions>>
```

+

> NOTE: Placeholder fields within an Osquery Response Action query will only work if the rule has generated alerts.

Reviewer 2: Not sure I understand the purpose of this note. I could be missing something, but why would someone expect a response action to run if it's parent rule hadn't generated an alert?

* <<invest-guide-run-osquery,Investigation guides using Osquery queries>> 

[float]
[[placeholder-field-syntax]]
=== Placeholder field syntax and requirements

> Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{}}`). You can use any field within an event or alert document as a placeholder field. 

Reviewer 3: All those brackets are a little hard to parse visually. How about something like:

```suggestion
Placeholder fields use http://mustache.github.io/[mustache syntax] and must be wrapped in double curly brackets (`{{example.field}}`). You can use any field within an event or alert document as a placeholder field. 
```

> Queries with placeholder fields _must_ be run against an alert or event. If a placeholder field can't find an alert or event value, the query status will be `Skipped`. 

Reviewer 2: ```suggestion
Queries with placeholder fields can only run against alerts or events. Otherwise, they will lack the necessary values and be `Skipped`. 
```

Author: @benironside I'm fairly certain that the second sentence needs to be kept as is. If an Osquery query is using a placeholder field to retrieve a field that doesn't exist on the alert or event, the query will fail regardless of where it was ran against. @szwarckonrad @tomsonpl can you verify what I said is correct? 

@szwarckonrad @tomsonpl - On a related note, during my testing, I've seen the `error` status returned for failed queries (i.e. queries using placeholder fields to pass in values for fields that don't exist on an alert or event). Is this correct? Should I replace `Skipped` with `error`?

Reviewer 4: Hey @nastasha-solomon , @benironside - to me both of your version seem the same (correct) :P 

Whenever we use the {{ data }} dynamic params syntax within context of Alert or Event (we need its data) - we try to replace the parameter with the data, and if we fail (eg. there is no data held under the key we provided) - we return an `error` saying "This query hasn't been called due to parameter used and its value not found in the alert." which means that in the end we didn't run this osquery on the host (not sending action to fleet). To me this can safely be referred to as something like "The query was skipped" 

I am sorry as this might sound confusing, but I'll be happy to discuss this further if that would help 👍

Author: Ah, ok thanks for the additional information about the error message and how Fleet perceives (or doesn't perceive?) the "failed" query. With this in mind, I agree that @benironside's edit makes sense to use. Thanks, both!

[float]
[[placeholder-field-example]]
==== Example query with a placeholder field

In the following query, the `{{host.name}}` placeholder field is used: 

`SELECT * FROM os_version WHERE name = `{{host.name}}``

> When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.name}}` placeholder field. 

Reviewer 2: very clear example, great illustration

Author: ```suggestion
When you run the query, the value that's stored in the alert or event's `host.name` field will be passed into the `{{host.os.name}}` placeholder field. 
```
